paddle/fluid/operators/math/matrix_bit_code.h

```diff
@@ -104,69 +124,166 @@ struct SimpleCode {
    * Binary classification path is the suffixes of encoding, thus leave out the
    * left most bit in calc_bit.
    */
-  inline size_t calc_index(int bit) const { return (c_ >> (bit + 1)) - 1; }
-  inline bool calc_bit(int bit) const { return c_ & (1 << bit); }
-  inline int get_length() const { return FindLastSet(c_) - 1; }
+  size_t calc_index(int bit) const { return (c_ >> (bit + 1)) - 1; }
+  bool calc_bit(int bit) const { return c_ & (1 << bit); }
+  int get_length() const { return FindLastSet(c_) - 1; }
 
  private:
   size_t c_;
 };
 
-struct SimpleCodeTable {
-  explicit SimpleCodeTable(size_t num_classes) : num_classes_(num_classes) {}
-  SimpleCode operator()(size_t code) const {
-    return SimpleCode(code, num_classes_);
+template <typename R>
+class CustomCode : public Code {
+ public:
+  CustomCode(const framework::LoDTensor* ptable,
+             const framework::LoDTensor* pcode, const int64_t* ids,
+             const int index)
+      : ptable_(ptable), pcode_(pcode), ids_(ids), index_(index) {}
+  /**
+   * Here the id of root shoud be 1 rather than 0, thus the encoding of class c
+   * is `c + num_classes` and all siblings can get the same weight indice using
+   * prefixes.
+   * Weight index is the prefixes of encoding, thus leave out the right most
+   * bit in calc_index.
+   * Binary classification path is the suffixes of encoding, thus leave out the
+   * left most bit in calc_bit.
+   */
+  size_t calc_index(int bit) const {
+    return ptable_
+        ->data<R>()[index_ * static_cast<int>(ptable_->dims()[1]) + bit];
+  }
+  bool calc_bit(int bit) const {
+    return pcode_
+        ->data<R>()[index_ * static_cast<int>(ptable_->dims()[1]) + bit];
+  }
+  int get_length() const {
+    int length = 0;
+
+    for (int i = 0; i < static_cast<int>(ptable_->dims()[1]); i++) {
+      if (ptable_->data<R>()[index_ * static_cast<int>(ptable_->dims()[1]) +
+                             i] >= 0) {
+        length++;
```

Comment:
It seems that `CustomCode` only access index row of a tensor, why not slice the tensor first rather than calculate `index*dims[1]` time by time?

Reply:
changed, thx